File: lib/disdex-pending-exposure-registry.ts
import { lstat, mkdir, readFile, rename, unlink, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";

export const PENDING_EXPOSURE_SCHEMA = "disdex-pending-exposure/v1" as const;
export type PendingExposureSleeve = "CRYPTO" | "STOCK";
export type PendingExposureStatus = "PENDING" | "SUBMITTED" | "UNKNOWN" | "RELEASED";

export interface PendingExposureEntry {
    reservationId: string;
    strategyId: string;
    sleeve: PendingExposureSleeve;
    symbol: string;
    side: "LONG" | "SHORT" | "FLAT";
    gross: number;
    notionalUsd: number;
    status: PendingExposureStatus;
    createdAt: number;
    updatedAt: number;
    runtimeSha?: string;
    idempotencyKey?: string;
}

export interface PendingExposureRegistry {
    schema: typeof PENDING_EXPOSURE_SCHEMA;
    accountScope: string;
    updatedAt: number;
    entries: PendingExposureEntry[];
}

export interface PendingExposureAggregate {
    cryptoGross: number;
    stockGross: number;
    byStrategyGross: Record<string, number>;
}

const ACTIVE_STATUSES = new Set<PendingExposureStatus>(["PENDING", "SUBMITTED", "UNKNOWN"]);

function finiteNonNegative(value: unknown, name: string) {
    const number = Number(value);
    if (!Number.isFinite(number) || number < 0) throw new Error(`PENDING_EXPOSURE_${name}_INVALID`);
    return number;
}

function finitePositive(value: unknown, name: string) {
    const number = Number(value);
    if (!Number.isFinite(number) || number <= 0) throw new Error(`PENDING_EXPOSURE_${name}_INVALID`);
    return number;
}

function expectedSleeve(strategyId: string): PendingExposureSleeve {
    const owner = String(strategyId || "").trim().toUpperCase();
    if (owner === "V11_EQ" || owner === "V50_POST_OPEN_BASIS" || owner === "V52" || owner.includes("V52")) return "STOCK";
    if (
        owner === "V12"
        || owner.includes("V12_")
        || owner.includes("PENGU")
        || owner.includes("FET_BRK48")
        || owner.includes("QUALITY102")
        || owner === "HYPE_LONG"
        || owner === "ZEC_LONG"
    ) return "CRYPTO";
    throw new Error(`PENDING_EXPOSURE_OWNER_UNKNOWN:${strategyId || "EMPTY"}`);
}

function normalizeEntry(raw: unknown): PendingExposureEntry {
    if (!raw || typeof raw !== "object") throw new Error("PENDING_EXPOSURE_ENTRY_INVALID");
    const row = raw as Partial<PendingExposureEntry>;
    const reservationId = String(row.reservationId || "").trim();
    const strategyId = String(row.strategyId || "").trim();
    const sleeve = String(row.sleeve || "").trim().toUpperCase() as PendingExposureSleeve;
    const symbol = String(row.symbol || "").trim().toUpperCase();
    const side = String(row.side || "").trim().toUpperCase() as PendingExposureEntry["side"];
    const status = String(row.status || "").trim().toUpperCase() as PendingExposureStatus;
    if (!reservationId || !strategyId || !symbol) throw new Error("PENDING_EXPOSURE_ENTRY_IDENTITY_INVALID");
    if (sleeve !== "CRYPTO" && sleeve !== "STOCK") throw new Error("PENDING_EXPOSURE_SLEEVE_INVALID");
    if (side !== "LONG" && side !== "SHORT" && side !== "FLAT") throw new Error("PENDING_EXPOSURE_SIDE_INVALID");
    if (!ACTIVE_STATUSES.has(status) && status !== "RELEASED") throw new Error("PENDING_EXPOSURE_STATUS_INVALID");
    if (expectedSleeve(strategyId) !== sleeve) throw new Error(`PENDING_EXPOSURE_OWNER_SLEEVE_MISMATCH:${strategyId}`);
    return {
        reservationId,
        strategyId,
        sleeve,
        symbol,
        side,
        gross: finiteNonNegative(row.gross, "GROSS"),
        notionalUsd: finiteNonNegative(row.notionalUsd, "NOTIONAL"),
        status,
        createdAt: finitePositive(row.createdAt, "CREATED_AT"),
        updatedAt: finitePositive(row.updatedAt, "UPDATED_AT"),
        ...(row.runtimeSha ? { runtimeSha: String(row.runtimeSha).trim().toLowerCase() } : {}),
        ...(row.idempotencyKey ? { idempotencyKey: String(row.idempotencyKey).trim() } : {}),
    };
}

export function normalizePendingExposureRegistry(raw: unknown): PendingExposureRegistry {
    if (!raw || typeof raw !== "object") throw new Error("PENDING_EXPOSURE_REGISTRY_MALFORMED");
    const value = raw as Partial<PendingExposureRegistry>;
    if (value.schema !== PENDING_EXPOSURE_SCHEMA) throw new Error("PENDING_EXPOSURE_REGISTRY_MALFORMED");
    if (!Array.isArray(value.entries)) throw new Error("PENDING_EXPOSURE_REGISTRY_MALFORMED");
    const accountScope = String(value.accountScope || "").trim();
    if (!accountScope) throw new Error("PENDING_EXPOSURE_ACCOUNT_SCOPE_INVALID");
    const entries = value.entries.map(normalizeEntry);
    const ids = new Set<string>();
    for (const entry of entries) {
        if (ids.has(entry.reservationId)) throw new Error(`PENDING_EXPOSURE_DUPLICATE:${entry.reservationId}`);
        ids.add(entry.reservationId);
    }
    return {
        schema: PENDING_EXPOSURE_SCHEMA,
        accountScope,
        updatedAt: finitePositive(value.updatedAt, "REGISTRY_UPDATED_AT"),
        entries,
    };
}

export function emptyPendingExposureRegistry(now = Date.now(), accountScope = "ASTER_FUTURES"): PendingExposureRegistry {
    return { schema: PENDING_EXPOSURE_SCHEMA, accountScope, updatedAt: now, entries: [] };
}

export function aggregatePendingExposure(registry: PendingExposureRegistry, options?: { excludeStrategyIds?: string[] }): PendingExposureAggregate {
    const result: PendingExposureAggregate = { cryptoGross: 0, stockGross: 0, byStrategyGross: {} };
    const excluded = new Set((options?.excludeStrategyIds || []).map((value) => String(value).trim().toUpperCase()));
    for (const entry of registry.entries) {
        if (!ACTIVE_STATUSES.has(entry.status)) continue;
        if (excluded.has(entry.strategyId.toUpperCase())) continue;
        if (entry.sleeve === "CRYPTO") result.cryptoGross += entry.gross;
        else result.stockGross += entry.gross;
        result.byStrategyGross[entry.strategyId] = (result.byStrategyGross[entry.strategyId] || 0) + entry.gross;
    }
    return result;
}

export function pendingExposurePath(path = process.env.DISDEX_PENDING_EXPOSURE_REGISTRY_PATH): string {
    return resolve(path || "/var/lib/disdex/shared/pending-exposure.json");
}

async function assertSafePath(path: string, allowMissing = true) {
    try {
        const stats = await lstat(path);
        if (stats.isSymbolicLink()) throw new Error("PENDING_EXPOSURE_PATH_SYMLINK");
        if (!stats.isFile()) throw new Error("PENDING_EXPOSURE_PATH_NOT_REGULAR_FILE");
    } catch (error) {
        if (allowMissing && (error as NodeJS.ErrnoException)?.code === "ENOENT") return;
        throw error;
    }
}

export async function readPendingExposureRegistry(path = pendingExposurePath()): Promise<PendingExposureRegistry> {
    await assertSafePath(path);
    try {
        return normalizePendingExposureRegistry(JSON.parse(await readFile(path, "utf8")));
    } catch (error) {
        if ((error as NodeJS.ErrnoException)?.code === "ENOENT") return emptyPendingExposureRegistry();
        throw error;
    }
}

async function atomicWrite(path: string, registry: PendingExposureRegistry) {
    await assertSafePath(path);
    await mkdir(dirname(path), { recursive: true });
    const temporary = `${path}.${process.pid}.${Date.now()}.tmp`;
    try {
        await writeFile(temporary, `${JSON.stringify(registry, null, 2)}\n`, { encoding: "utf8", mode: 0o600 });
        await rename(temporary, path);
    } catch (error) {
        await unlink(temporary).catch(() => undefined);
        throw error;
    }
}

export async function upsertPendingExposure(
    entry: Omit<PendingExposureEntry, "status" | "updatedAt"> & { status?: PendingExposureStatus; updatedAt?: number },
    path = pendingExposurePath(),
) {
    const current = await readPendingExposureRegistry(path);
    const normalized = normalizeEntry({ ...entry, status: entry.status || "PENDING", updatedAt: entry.updatedAt || Date.now() });
    const next = normalizePendingExposureRegistry({
        ...current,
        updatedAt: Date.now(),
        entries: [...current.entries.filter((row) => row.reservationId !== normalized.reservationId), normalized],
    });
    await atomicWrite(path, next);
    return normalized;
}

export async function releasePendingExposure(reservationId: string, path = pendingExposurePath()) {
    const current = await readPendingExposureRegistry(path);
    const id = String(reservationId || "").trim();
    if (!id) throw new Error("PENDING_EXPOSURE_RESERVATION_ID_INVALID");
    const found = current.entries.some((row) => row.reservationId === id);
    if (!found) return false;
    const next = normalizePendingExposureRegistry({
        ...current,
        updatedAt: Date.now(),
        entries: current.entries.map((row) => row.reservationId === id ? { ...row, status: "RELEASED", updatedAt: Date.now() } : row),
    });
    await atomicWrite(path, next);
    return true;
}

File: scripts/disdex_pending_exposure_registry.py
"""Cross-language durable pending exposure registry.

The registry is deliberately fail-closed. Active rows survive a runner crash
and continue to reserve Gross until a reconciliation path explicitly releases
the row. It is a coordination record only and never submits or cancels orders.
"""
from __future__ import annotations

import json
import os
import tempfile
import time
from pathlib import Path
from typing import Any

SCHEMA = "disdex-pending-exposure/v1"
ACTIVE_STATUSES = {"PENDING", "SUBMITTED", "UNKNOWN"}


def default_path() -> Path:
    return Path(os.getenv("DISDEX_PENDING_EXPOSURE_REGISTRY_PATH", "/var/lib/disdex/shared/pending-exposure.json")).resolve()


def _now_ms() -> int:
    return int(time.time() * 1000)


def _expected_sleeve(strategy_id: str) -> str:
    owner = str(strategy_id or "").strip().upper()
    if owner in {"V11_EQ", "V50_POST_OPEN_BASIS", "V52"} or "V52" in owner:
        return "STOCK"
    if owner == "V12" or "V12_" in owner or "PENGU" in owner or "FET_BRK48" in owner or "QUALITY102" in owner:
        return "CRYPTO"
    raise RuntimeError(f"PENDING_EXPOSURE_OWNER_UNKNOWN:{strategy_id or 'EMPTY'}")


def _validate_entry(row: Any) -> dict[str, Any]:
    if not isinstance(row, dict):
        raise RuntimeError("PENDING_EXPOSURE_ENTRY_INVALID")
    strategy_id = str(row.get("strategyId", "")).strip()
    sleeve = str(row.get("sleeve", "")).strip().upper()
    if not strategy_id or sleeve not in {"CRYPTO", "STOCK"}:
        raise RuntimeError("PENDING_EXPOSURE_ENTRY_IDENTITY_INVALID")
    if _expected_sleeve(strategy_id) != sleeve:
        raise RuntimeError(f"PENDING_EXPOSURE_OWNER_SLEEVE_MISMATCH:{strategy_id}")
    status = str(row.get("status", "")).strip().upper()
    if status not in ACTIVE_STATUSES | {"RELEASED"}:
        raise RuntimeError("PENDING_EXPOSURE_STATUS_INVALID")
    for key in ("gross", "notionalUsd"):
        value = float(row.get(key, 0))
        if value < 0:
            raise RuntimeError(f"PENDING_EXPOSURE_{key.upper()}_INVALID")
    for key in ("createdAt", "updatedAt"):
        value = int(row.get(key, 0))
        if value <= 0:
            raise RuntimeError(f"PENDING_EXPOSURE_{key.upper()}_INVALID")
    return dict(row, strategyId=strategy_id, sleeve=sleeve, status=status)


def empty_registry(now_ms: int | None = None, account_scope: str = "ASTER_FUTURES") -> dict[str, Any]:
    return {"schema": SCHEMA, "accountScope": account_scope, "updatedAt": now_ms or _now_ms(), "entries": []}


def read_registry(path: str | Path | None = None) -> dict[str, Any]:
    target = Path(path or default_path()).resolve()
    if not target.exists():
        return empty_registry()
    if target.is_symlink() or not target.is_file():
        raise RuntimeError("PENDING_EXPOSURE_PATH_NOT_REGULAR_FILE")
    try:
        value = json.loads(target.read_text(encoding="utf-8"))
    except Exception as error:
        raise RuntimeError("PENDING_EXPOSURE_REGISTRY_MALFORMED") from error
    if not isinstance(value, dict) or value.get("schema") != SCHEMA or not isinstance(value.get("entries"), list):
        raise RuntimeError("PENDING_EXPOSURE_REGISTRY_MALFORMED")
    entries = [_validate_entry(row) for row in value["entries"]]
    if len({row["reservationId"] for row in entries}) != len(entries):
        raise RuntimeError("PENDING_EXPOSURE_DUPLICATE")
    return dict(value, entries=entries)


def _atomic_write(path: Path, value: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temp = tempfile.mkstemp(prefix=f"{path.name}.", suffix=".tmp", dir=path.parent)
    try:
        os.fchmod(fd, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            json.dump(value, handle, indent=2, sort_keys=True)
            handle.write("\n")
        os.replace(temp, path)
    finally:
        try:
            os.unlink(temp)
        except FileNotFoundError:
            pass


def upsert_pending(entry: dict[str, Any], path: str | Path | None = None) -> dict[str, Any]:
    target = Path(path or default_path()).resolve()
    current = read_registry(target)
    normalized = _validate_entry(dict(entry, status=entry.get("status", "PENDING"), updatedAt=entry.get("updatedAt", _now_ms())))
    next_entries = [row for row in current["entries"] if row.get("reservationId") != normalized["reservationId"]]
    next_entries.append(normalized)
    _atomic_write(target, dict(current, updatedAt=_now_ms(), entries=next_entries))
    return normalized


def release_pending(reservation_id: str, path: str | Path | None = None) -> bool:
    target = Path(path or default_path()).resolve()
    current = read_registry(target)
    found = False
    next_entries = []
    for row in current["entries"]:
        if row.get("reservationId") == reservation_id:
            found = True
            next_entries.append(dict(row, status="RELEASED", updatedAt=_now_ms()))
        else:
            next_entries.append(row)
    if found:
        _atomic_write(target, dict(current, updatedAt=_now_ms(), entries=next_entries))
    return found


def aggregate_pending(path: str | Path | None = None) -> dict[str, Any]:
    result: dict[str, Any] = {"cryptoGross": 0.0, "stockGross": 0.0, "byStrategyGross": {}}
    for row in read_registry(path)["entries"]:
        if row["status"] not in ACTIVE_STATUSES:
            continue
        gross = float(row["gross"])
        result["cryptoGross" if row["sleeve"] == "CRYPTO" else "stockGross"] += gross
        result["byStrategyGross"][row["strategyId"]] = result["byStrategyGross"].get(row["strategyId"], 0.0) + gross
    return result
